lib/seattleflu/db/cli/command/etl/presence_absence.py

```diff
@@ -0,0 +1,397 @@
+    return control_status == "PositiveControl"
+
+
+def find_or_create_sample(db: DatabaseSession, identifier: str,
+                          details: dict) -> Any:
+    """
+    Select sample by *identifier*, or insert it if it doesn't exist.
+
+    TODO eventually we'll need to add the encounter_id once the relationship
+    between encounter_id and sample_id exists in the db.
+    """
+    LOG.debug(f"Looking up sample «{identifier}»")
+
+    sample = db.fetch_row("""
+        select sample_id as id, identifier
+          from warehouse.sample
+         where identifier = %s
+        """, (identifier,))
+
+    if sample:
+        LOG.info(f"Found sample {sample.id} «{sample.identifier}»")
+    else:
+        LOG.debug(f"Sample «{identifier}» not found, adding")
+
+        data = {
+            "identifier": identifier,
+            "details": Json(details)
+            # "encounter_id"
+        }
+
+        sample = db.fetch_row("""
+            insert into warehouse.sample (
+                identifier, 
+                details)
+                
+                values (
+                    %(identifier)s, 
+                    %(details)s)
+                    
+            returning sample_id as id, identifier
+            """, data)
+
+        LOG.info(f"Created sample {sample.id} «{sample.identifier}»")
+
+    return sample
+
+def sample_identifier(db: DatabaseSession, barcode: str) -> str:
+    """
+    Find corresponding UUID for scanned sample barcode within
+    warehouse.identifier.
+
+    Returns barcode if it starts with "SCH-" as this indicates it 
+    is a control sample, and therefore does not have a UUID.
+    
+    TODO determine course of action if barcode not found
+    within warehouse.identifier.
+    """
+
+    LOG.debug(f"Looking up sample barcode {barcode} to find UUID")
+
+    uuid = db.fetch_row("""
+        select uuid, name
+          from warehouse.identifier
+          join warehouse.identifier_set using (identifier_set_id)
+         where barcode = %s
+        """, (barcode,))
+    if uuid:
+        #Check Identifier_Set
+        if uuid.name == "samples":
+            LOG.info(f"Found sample UUID {uuid.uuid}")
+            return uuid.uuid
+        else:
+            raise IncorrectIdentifierSetError(f"Found sample UUID {uuid.uuid}, but UUID is of the incorrect identifier set «{uuid.name}»")
+    else:
+        LOG.warning(f"No corresponding UUID found for barcode {barcode}")
+        return barcode
+
+
+def sample_details(document: dict) -> dict:
+    """
+    Describe sample details in a simple data structure designed to be used
+    from SQL.
+    """
+    # sampleproceedtosequencing
+    # initialproceedtosequencingcall
+    return { 
+
+        "sampleComment": document['sampleComment'],
+        "initialSequenceCall": document['initialProceedToSequencingCall'],
+        "proceedToSequence": document["sampleProceedToSequencing"]
+    }
```

Comment:
Let's standardize key names on `snake_case`. This is the convention used by more or less all our software in the lab, and what we're using in the database as well. (One exception is the upper camel-case questionnaire response ids, which we don't control.)

Reply:
We might also rename these keys to something less repetitive and more consistent than what we receive. Maybe:

* `comment`
* `initial_sequencing_call`
* `final_sequencing_call`

Reply:
See my other comment here too :point_up: